Objects/rangeobject.c

```diff
@@ -816,13 +816,29 @@ PyTypeObject PyRange_Type = {
 static PyObject *
 rangeiter_next(_PyRangeIterObject *r)
 {
+#ifdef Py_GIL_DISABLED
+    // r->step is readonly attribute, so we can assume it is not changed.
+    long step = r->step;
+    do {
+        long len = _Py_atomic_load_long_relaxed(&r->len);
+        if (len <= 0) {
+            return NULL;
+        }
+        long result = _Py_atomic_load_long_relaxed(&r->start);
```

Comment:
What if we have a range object with `len=1` and many threads simultaneously at this point? The check on `len` has already been done, so the threads each increment the value of `start` and we end up with a value for `reault` larger than `end`.

Reply:
Maybe fail at _Py_atomic_compare_exchange_long?

Reply:
Ah.. yes understood what you pointed out.

Reply:
There will be a slight timing issue. Then yes, let's just use the critical section that @Fidget-Spinner also preferred.

Reply:
To prepare for the overcommitted issue, we can validate the value of the result by comparing the end value.
But let's just use critical section approach.